### PR TITLE
feat: update Dockerfile, manifest and setup files

### PR DESCRIPTION
**Pull request by alaeddine-13**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,7 +1,7 @@
-FROM jinaai/jina:2.0.0rc7
+FROM jinaai/jina:2.0.5
```
Comment (nan-wang) [inline Dockerfile]: Let's update it to the latest release
```suggestion
FROM jinaai/jina:2.0.7
```

**Review by nan-wang — DISMISSED** [review]
Please add some tests

**Comment by alaeddine-13** [comment]
@nan-wang @JoanFM after receivign assistance from david, I'm changing the encoding method by avoiding the batch normalization step. So I'll convert this PR to draft

File: audio_clip_encoder.py [hunk 2]
```diff
@@ -55,10 +53,8 @@ def _create_embeddings(self, filtered_docs: Iterable):
         """Update the documents with the embeddings generated by AudioCLIP"""
 
         for d in filtered_docs:
-            audio = [d.blob, deepcopy(d.blob)]
-            audio = torch.stack([self.audio_transforms(track.reshape(1, -1)) for track in audio])
+            track = d.blob
+            audio = torch.stack([self.audio_transforms(track.reshape(1, -1))])
```
Comment (JoanFM) [inline audio_clip_encoder.py]: are we sure `audio` cannot work in `batches`?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: audio_clip_encoder.py (hunk 2 above)
Comment (alaeddine-13) [reply audio_clip_encoder.py]: so you want to stack all document blobs together and encode them at once ?

Comment (JoanFM) [reply audio_clip_encoder.py]: not all, but work in batches, like with other executors

Comment (alaeddine-13) [reply audio_clip_encoder.py]: should be done

File: Dockerfile (hunk 1 above)
Comment (alaeddine-13) [reply Dockerfile]: updated

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -9,4 +9,7 @@ jobs:
       - name: Checkout
         uses: actions/checkout@v2
       - id: tests
-        run: ./scripts/tests.sh
\ No newline at end of file
+        run: |
+          sudo apt-get install libsndfile-dev ffmpeg
```
Comment (JoanFM) [inline .github/workflows/ci.yml:13]: why is this needed here and not in the `Dockerfile`?

File: audio_clip_encoder.py [hunk 4]
```diff
@@ -30,36 +29,35 @@ def __init__(self,
...
-            ((audio_features, _, _), _), _ = self.aclp(audio=audio)
-            audio_features = audio_features[0]
-            audio_features = audio_features / torch.linalg.norm(audio_features, dim=-1, keepdim=True)
+    def _create_embeddings(self, document_batches_generator: Iterable):
+        """Update the documents with the embeddings generated by AudioCLIP"""
 
-            d.embedding = audio_features.numpy()
+        for document_batch in document_batches_generator:
+            audio = torch.stack([self.audio_transforms(d.blob.reshape(1, -1)) for d in document_batch])
+            ((embedding_batch, _, _), _), _ = self.aclp(audio=audio)
```
Comment (JoanFM) [inline audio_clip_encoder.py]: u should not call `self.aclp` but `self.aclp.encode_audio()`

File: audio_clip_encoder.py [hunk 5]
```diff
@@ -30,36 +29,35 @@ def __init__(self,
...
+            embedding_batch = embedding_batch / torch.linalg.norm(embedding_batch, dim=-1, keepdim=True)
```
Comment (JoanFM) [inline audio_clip_encoder.py]: why do we do this?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: model/audioclip.py [hunk 6]
```diff
@@ -88,6 +88,8 @@ def __init__(self,
             pretrained=False
         )
 
+        self.audio.eval()
```
Comment (JoanFM) [inline model/audioclip.py]: this should not be here

File: audio_clip_encoder.py [hunk 7]
```diff
@@ -30,36 +29,35 @@ def __init__(self,
         self.aclp = AudioCLIP(pretrained=model_path)
```
Comment (JoanFM) [inline audio_clip_encoder.py:28]: here u should set `self.aclp.eval()`

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 3 above)
Comment (alaeddine-13) [reply .github/workflows/ci.yml:13]: loading audio with librosa is performed during unit tests therefore installing libsndfile-dev and ffmpeg is needed.
However, been thinking that the executor itself doesn't need to load the images so it's not need so that the executor works.
What do you think ?

Comment (JoanFM) [reply .github/workflows/ci.yml:13]: good point

File: audio_clip_encoder.py (hunk 5 above)
Comment (alaeddine-13) [reply audio_clip_encoder.py]: these are preprocessing and postprocessing steps performed in the AudioCLIP's demo in order to get embeddings useful for query: https://github.com/AndreyGuzhov/AudioCLIP/blob/master/demo/AudioCLIP.ipynb

File: audio_clip_encoder.py (hunk 7 above)
Comment (alaeddine-13) [reply audio_clip_encoder.py:28]: should I override method AudioCLIP:eval so that it calls self.audio.eval() ?

Comment (JoanFM) [reply audio_clip_encoder.py:28]: no, `nn.Module` eval should know that this is done. But anyway, u should change it from the outside. You should not change the code that is not ours because in other versions it may change

Comment (alaeddine-13) [reply audio_clip_encoder.py:28]: or just use self.aclp.audio.eval() ?

File: audio_clip_encoder.py (hunk 5 above)
Comment (JoanFM) [reply audio_clip_encoder.py]: this is not used in the `CLIPTextEncoder` so it cannot be here for now.

See that this is postprocess needed to have a classification layer, but to work on the embedding space it can be good.

For now, we need to remove it and we can discuss if we need to add this `normalization` step in the other `CLIP` executors

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: tests/integration/test_exec.py [hunk 8]
```diff
@@ -1,12 +1,29 @@
...
+
+def test_embedding_exists():
+
+    x_audio, sample_rate = librosa.load(os.path.join(cur_dir, '../data/sample.mp3'))
+    doc = DocumentArray([Document(blob=x_audio)])
+
+    with Flow.load_config(os.path.join(cur_dir, 'flow.yml')) as f:
+        responses = f.post(on='index', inputs=doc, return_results=True)
+
+    assert responses[0].docs[0].embedding is not None
```
Comment (JoanFM) [inline tests/integration/test_exec.py:29]: assert properly the shape at least

**Review by JoanFM — CHANGES_REQUESTED** [review]
we need many more tests with different traversal paths and options

**Comment by alaeddine-13** [comment]
> we need many more tests with different traversal paths and options

I added tests with traversal paths

File: audio_clip_encoder.py [hunk 9]
```diff
@@ -52,14 +52,10 @@ def _get_input_data(self, docs: DocumentArray, parameters: dict):
...
-            # Vggish broadcasts across different length audios, not batches
-            audio = [d.blob, deepcopy(d.blob)]
-            audio = torch.stack([self.audio_transforms(track.reshape(1, -1)) for track in audio])
-            ((audio_features, _, _), _), _ = self.aclp(audio=audio)
-            audio_features = audio_features[0]
-            audio_features = audio_features / torch.linalg.norm(audio_features, dim=-1, keepdim=True)
-
-            d.embedding = audio_features.numpy()
+            track = d.blob
+            audio = torch.stack([self.audio_transforms(track.reshape(1, -1))])
```
Comment (JoanFM) [inline audio_clip_encoder.py]: synthetize this in a single line

Comment (JoanFM) [inline audio_clip_encoder.py]: why is the reshape needed or the `stack` is it weird to use `stack` when a signle `blob` is used

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: audio_clip_encoder.py [hunk 10]
```diff
@@ -52,14 +51,9 @@ def _get_input_data(self, docs: DocumentArray, parameters: dict):
...
-            audio = [d.blob, deepcopy(d.blob)]
-            audio = torch.stack([self.audio_transforms(track.reshape(1, -1)) for track in audio])
-            ((audio_features, _, _), _), _ = self.aclp(audio=audio)
-            audio_features = audio_features[0]
-            audio_features = audio_features / torch.linalg.norm(audio_features, dim=-1, keepdim=True)
-
-            d.embedding = audio_features.numpy()
+            audio = torch.Tensor(d.blob).unsqueeze(0)
+            embedding = self.aclp.encode_audio(audio=audio)[0]
+            d.embedding = embedding.numpy()
```
Comment (JoanFM) [inline audio_clip_encoder.py]:
```suggestion
            d.embedding = embedding.cpu().numpy()
```

File: config.yml [hunk 11]
```diff
@@ -1,4 +1,3 @@
 jtype: AudioCLIPEncoder
 metas:
-  py_modules:
-    - audio_clip_encoder.py
+  py_modules: [audio_clip_encoder.py]
```
Comment (JoanFM) [inline config.yml]: keep as it was

File: tests/integration/test_exec.py [hunk 12]
```diff
@@ -1,12 +1,93 @@
...
+    audio1, _ = librosa.load(os.path.join(cur_dir, '../data/sample.mp3'))
+    audio2, _ = librosa.load(os.path.join(cur_dir, '../data/sample.wav'))
+    doc = DocumentArray([Document(blob=audio1), Document(blob=audio2)])
+
+    with Flow.load_config(os.path.join(cur_dir, 'flow.yml')) as f:
+        responses = f.post(on='index', inputs=doc, return_results=True)
+
+    assert responses[0].docs[0].embedding.shape == (1024, ) and responses[0].docs[1].embedding.shape == (1024, )
+
+def test_traversal_paths():
```
Comment (JoanFM) [inline tests/integration/test_exec.py]: these tests should go to `unit tests`.

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: tests/integration/test_exec.py (hunk 12 above)
Comment (alaeddine-13) [reply tests/integration/test_exec.py]: are you referring to test_traversal_paths or the test_many_documents ?

Comment (JoanFM) [reply tests/integration/test_exec.py]: specially traversal paths and potentially also the other.

Comment (alaeddine-13) [reply tests/integration/test_exec.py]: applied the changes

File: tests/unit/test_exec.py [hunk 13]
```diff
@@ -1,9 +1,76 @@
...
+            ]
+        )
+    ])
+
+    model = AudioCLIPEncoder(default_traversal_paths=['c'])
+    model.encode(docs, parameters={})
+    model.encode(docs, parameters={'traversal_paths': ['cc']})
+    for path, count in [['r', 0], ['c', 6], ['cc', 3]]:
+        embeddings = DocumentArray(docs).traverse_flat([path]).get_attributes('embedding')
+        assert all(
```
Comment (JoanFM) [inline tests/unit/test_exec.py:74]: separate into 2 asserts, is easier to read

File: tests/unit/test_exec.py [hunk 14]
```diff
@@ -1,9 +1,76 @@
...
+def test_many_documents():
+
+    audio1, _ = librosa.load(os.path.join(cur_dir, '../data/sample.mp3'))
+    audio2, _ = librosa.load(os.path.join(cur_dir, '../data/sample.wav'))
+    docs = DocumentArray([Document(blob=audio1), Document(blob=audio2)])
+
+    model = AudioCLIPEncoder()
+    model.encode(docs, parameters={})
+
+    assert docs[0].embedding.shape == (1024, ) and docs[1].embedding.shape == (1024, )
```
Comment (JoanFM) [inline tests/unit/test_exec.py]: separate in multiple asserts

File: tests/integration/test_exec.py [hunk 15]
```diff
@@ -1,12 +1,29 @@
...
+
+def test_embedding_exists():
+
+    x_audio, _ = librosa.load(os.path.join(cur_dir, '../data/sample.mp3'))
+    doc = DocumentArray([Document(blob=x_audio)])
+
+    with Flow.load_config(os.path.join(cur_dir, 'flow.yml')) as f:
+        responses = f.post(on='index', inputs=doc, return_results=True)
+
+    assert responses[0].docs[0].embedding is not None and responses[0].docs[0].embedding.shape == (1024, )
```
Comment (JoanFM) [inline tests/integration/test_exec.py]: separate in multiple asserts

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: tests/unit/test_exec.py (hunk 13 above)
Comment (alaeddine-13) [reply tests/unit/test_exec.py:74]: done

File: tests/integration/test_exec.py (hunk 15 above)
Comment (alaeddine-13) [reply tests/integration/test_exec.py]: done

File: tests/unit/test_exec.py (hunk 14 above)
Comment (alaeddine-13) [reply tests/unit/test_exec.py]: done

**Review by JoanFM — APPROVED** [review]
